recipe-client-addon/lib/Sampling.jsm

```diff
@@ -57,25 +55,72 @@ function bufferToHex(buffer) {
   return hexCodes.join("");
 }
 
+function rangeBucketSample(inputHash, minBucket, maxBucket, bucketCount) {
+  const minHash = fractionToKey(minBucket / bucketCount);
+  const maxHash = fractionToKey(maxBucket / bucketCount);
+  return (minHash <= inputHash) && (inputHash < maxHash);
+}
+
+/**
+ * Return a promise for a truncated hash of `data`.
+ */
+const truncatedHash = Task.async(function* (data) {
+  const hasher = crypto.subtle;
+  const input = new TextEncoder("utf-8").encode(JSON.stringify(data));
+  const hash = yield hasher.digest("SHA-256", input);
+    // truncate hash to 12 characters (2^48), because the full hash is larger
+    // than JS can meaningfully represent as a number.
+  return bufferToHex(hash).slice(0, 12);
+});
+
 this.Sampling = {
-  stableSample(input, rate) {
-    const hasher = crypto.subtle;
-
-    return hasher.digest("SHA-256", new TextEncoder("utf-8").encode(JSON.stringify(input)))
-      .then(hash => {
-        // truncate hash to 12 characters (2^48)
-        const inputHash = bufferToHex(hash).slice(0, 12);
-        const samplePoint = fractionToKey(rate);
-
-        if (samplePoint.length !== 12 || inputHash.length !== 12) {
-          throw new Error("Unexpected hash length");
-        }
-
-        return inputHash < samplePoint;
-
-      })
-      .catch(error => {
-        log.error(`Error: ${error}`);
-      });
-  },
+  /**
+   * Sample by splitting the input into two buckets, one with a size (rate) and
+   * another with a size (1.0 - rate), and then check if the input's hash falls
+   * into the first bucket.
+   *
+   * @param {object} input Input to hash to determine the sample.
+   * @param {Number} rate Number between 0.0 and 1.0 to sample at. A value of
+   *                      0.25 returns true 25% of the time.
+   * @promises {boolean}    True if the input is in the sample.
```

Comment:
Same here RE: alignment. Actually, I dunno what the standard is for this: a hanging indent and per-line alignment, or all the descriptions lined up together?